client/control/reducers/FiltersReducer.js

```diff
@@ -0,0 +1,197 @@
+/**
+ * Store for tracking front-end 'filter list' status
+ *
+ * Exists as an array of objects denoting option groups,
+ * containing each set of options per group.
+ *
+ * Active/enabled filters have a `selected` property,
+ * which is mostly used later in the filter selectors
+ */
+
+import {
+  LOAD_FILTERS,
+  SET_FILTER,
+  SET_ALL_FILTERS,
+  SET_TEXT_FILTER,
+} from 'control/actions/FilterActions';
+
+import cloneArrayValues from 'client/utils/clone-array';
+
+function formatLabel(value) {
+  return value.charAt(0).toUpperCase() + value.slice(1);
+}
+
+function formatFilterOption(option) {
+  let label;
+  let value;
+
+  // string type = we were given just an option
+  // (no keyed value/label)
+  if (typeof option === 'string') {
+    value = option;
+    label = formatLabel(option);
+  // else if we get an object, then we
+  // can extract the key/value props
+  } else if (typeof option === 'object') {
+    label = option.value;
+    value = option.key;
+  }
+
+  return {
+    value,
+    label,
+  };
+}
+
+// Filters start out empty, as we need to load them from the API
+let initialState = [];
+
+function filtersReducer(state = initialState, action) {
+  let newState;
+  let textOptions;
+
+  switch (action.type) {
+    case LOAD_FILTERS: {
+      newState = [];
+      const { filters } = action;
+
+      for (const group in filters) {
+        if (!filters.hasOwnProperty(group)) {
+          break;
+        }
+        const newGroup = {
+          value: group,
+          label: formatLabel(group),
+          multiple: filters[group].length > 2,
+          options: cloneArrayValues(filters[group]).map(formatFilterOption),
+        };
+
+        newState.push(newGroup);
+      }
+
+      initialState = cloneArrayValues(newState);
+
+      break;
+    }
+
+    case SET_ALL_FILTERS: {
+      newState = cloneArrayValues(action.filters || initialState);
```

Comment:
Sounds good then. I'm glad the tests found the issues.

Do you think Immutable.js would be useful here? I've tried it both with and without, and I haven't really decided which I prefer. If you'd prefer switching to immutable, I'd support that.